core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -282,16 +284,26 @@ private void doExecute0(ClientRequestContext ctx, HttpRequestDuplicator rootReqD
             duplicateReq = rootReqDuplicator.duplicate(newHeaders.build());
         }
 
-        final ClientRequestContext derivedCtx;
+        final DefaultClientRequestContext derivedCtx;
         try {
-            derivedCtx = newDerivedContext(ctx, duplicateReq, ctx.rpcRequest(), initialAttempt);
+            derivedCtx = newDerivedContext(ctx, duplicateReq, ctx.rpcRequest(), initialAttempt)
+                    .as(DefaultClientRequestContext.class);
```

Comment:
I think it's always possible for `as()` to return `null`. Can't we just always downcast instead of implementing `Unwrappable`?

By the way, this starts to feel a lot like a hack 🥲

Reply:
I've reverted back to:
1) Checking if the context is an instance of `DefaultClientRequestContext`
2) If it is a `DefaultClientRequestContext`, downcast and possibly retry endpoint selection
3) If it isn't a `DefaultClientRequestContext`, then proceed with old logic

I think it is highly unlikely, but I think it's still possible that users use their own variant of `ClientRequestContext`, or wrap this with `ClientRequestContextWrapper`. The above logic has the advantage of:
1) Not breaking the implementation with an unexpected type
2) Less complexity, especially since it seems like `*RequestContextWrapper` isn't used anywhere in our codebase and may be deprecated

Reply:
Agreed. For the record, we had a chat about this some time ago and decided to introduce unwrapping in a separate future PR.